### ISSUE TITLE
🧮 Implement Gradient Accumulation and Computational Graph

### ISSUE THREAD

**Issue by sakshambedi** [sakshambedi/micrograd#13]
# 🧮 Implement Gradient Accumulation and Computational Graph

## 📋 Description

Implement a complete gradient accumulation system with computational graph construction and traversal. This is essential for enabling backpropagation through complex tensor operations and building neural networks.

## 🎯 Acceptance Criteria

- [ ] Implement computational graph construction during forward pass
- [ ] Add `backward()` method to Tensor class for gradient computation
- [ ] Implement gradient accumulation for tensors with `requires_grad=True`
- [ ] Support for retain_graph and create_graph options
- [ ] Add `zero_grad()` method to clear accumulated gradients
- [ ] Implement topological sorting for proper gradient computation order
- [ ] Handle in-place operations and their impact on gradients
- [ ] Add gradient clipping utilities
- [ ] Support for higher-order derivatives

### Current State
- Tensor has `grad` attribute: ✅ Implemented
- Operations save context: ✅ Partially implemented
- Backward methods: ❌ Missing (Issue #01)
- Graph traversal: ❌ Not implemented

### Core Components Needed

**Computational Graph Node**
```python
class GraphNode:
    def __init__(self, tensor, grad_fn=None, inputs=None):
        self.tensor = tensor
        self.grad_fn = grad_fn
        self.inputs = inputs or []
        self.next_functions = []
```

**Gradient Engine**
```python
class GradientEngine:
    @staticmethod
    def backward(tensor, grad_output=None, retain_graph=False):
        # Topological sort and gradient computation
        pass
```

**Tensor Integration**
```python
def backward(self, gradient=None, retain_graph=False, create_graph=False):
    """Compute gradients for this tensor and all dependencies"""
    pass

def zero_grad(self):
    """Clear accumulated gradients"""
    if self.grad is not None:
        self.grad = None
```

### Implementation Strategy

**Phase 1: Basic Graph Construction**
- [ ] Link tensors through operations
- [ ] Build dependency graph during forward pass
- [ ] Store necessary information for backward pass

**Phase 2: Gradient Computation**
- [ ] Implement topological sorting algorithm
- [ ] Add gradient accumulation logic
- [ ] Handle leaf vs non-leaf tensors

**Phase 3: Advanced Features**
- [ ] Support for retain_graph
- [ ] In-place operation handling
- [ ] Memory management for large graphs

## 🧪 Testing Requirements

- [ ] Test simple chain: a -> b -> c -> loss
- [ ] Test branching: a -> [b, c] -> d
- [ ] Test diamond pattern: a -> [b, c] -> d (where d uses both b and c)
- [ ] Test gradient accumulation across multiple backward passes
- [ ] Test memory management (graphs are freed when not retained)
- [ ] Test higher-order derivatives
- [ ] Verify gradients against numerical differentiation

## 📁 Files to Modify

- `grad/tensor.py` - Add backward() and zero_grad() methods
- `grad/autograd/engine.py` - New file for gradient engine
- `grad/autograd/graph.py` - New file for graph utilities
- `grad/autograd/function.py` - Enhance context management
- `tests/test_gradient_accumulation.py` - Comprehensive test suite

## 🔗 Dependencies

- Requires: Backward pass implementation (#01)
- Requires: Basic autograd operations (✅ forward passes done)
- Blocks: Neural network layers (#08)
- Blocks: Optimizers (#10)

### Basic Usage
```python
x = Tensor([2.0], requires_grad=True)
y = Tensor([3.0], requires_grad=True)

z = x * y + x
loss = z ** 2

loss.backward()

print(x.grad)  # Should contain accumulated gradient
print(y.grad)  # Should contain gradient w.r.t. y
```

### Multiple Backward Passes
```python
x = Tensor([1.0], requires_grad=True)

# First computation
y1 = x * 2
y1.backward()
print(x.grad)  # [2.0]

# Second computation (accumulate)
y2 = x * 3
y2.backward()
print(x.grad)  # [5.0] (accumulated: 2.0 + 3.0)

# Clear gradients
x.zero_grad()
print(x.grad)  # None or [0.0]
```

## 🚨 Edge Cases to Handle

- Backward on non-scalar tensors (should sum gradients)
- Cycles in computational graph (detect and error)
- Memory leaks from retained graphs
- Thread safety for gradient computation
- Gradient computation with mixed dtypes

## 🔗 Performance Considerations

- Minimize memory allocation during backward pass
- Efficient topological sorting implementation
- Lazy gradient computation where possible
- Memory pooling for gradient tensors

## 🏷️ Labels

- `enhancement`
- `autograd`
- `core-functionality`
- `high-priority`

## 📈 Priority

**High** - Essential for neural network training

## 🔗 Related Issues

- Depends on: Backward pass implementation (#01)
- Blocks: Optimizers (#10)
- Blocks: Neural network layers (#08)
- Related to: Memory management optimization
